### ISSUE TITLE
Erratic learning rate regarding FastSpeech2 (NoamAnnealing) and HiFi-GAN

### ISSUE THREAD

**Issue by DalHyun** [NVIDIA/NeMo#2384]
**Describe the bug**

Given the default configuration of fastspeech2 and HifiGAN, their implemented learning rates deviate severely from what's reported in their respective original papers.

**Steps/Code to reproduce bug**

Preprocess LJSpeech as specified by the official NeMo example, 
python fastspeech2.py (or hifigan.py)

**Expected behavior**
![expected](https://user-images.githubusercontent.com/37045193/122874113-95059880-d36d-11eb-8f87-7f35e06fb232.png)

**What is observed**
FastSpeech2
![observed](https://user-images.githubusercontent.com/37045193/122874437-f4fc3f00-d36d-11eb-8f53-5a76ca1e46f1.png)

HiFiGAN
![observed_hifigan](https://user-images.githubusercontent.com/37045193/122874450-f88fc600-d36d-11eb-85ab-a4e94f24462f.png)

We believe the error comes from the missing scaler, which needs to be 1000. given the default model dimension 256.
mult = self._normalize * min(step ** (-0.5), step * (self.warmup_steps ** (-1.5)))
mult = 1000. * self._normalize * min(step ** (-0.5), step * (self.warmup_steps ** (-1.5)))

**Comment by titu1994** [comment]
For the Noam Annealing scheduler, the config value of optim.lr acts as the linear scalar (you can explicitly set this to 1000 if required). You can see it in effect in ASR configs of Conformer which also use Noam Scheduler (smaller models use the optim.lr value of 5, larger models use 2 etc).

cc @blisc to see if this base LR scalar should be updated I the configs.

**Comment by DalHyun** [comment]
Thank you for your prompt answer.
Yes, as you said the optim.lr acts merely as the linear scalar, and it is not necessarily the "initial learning rate", which is a bit counter-intuitive. Things will be much easier if we can specify the initial and terminal weight level somewhere in the optim params.

  optim:
    name: adam
    betas: [0.9,0.98]
    lr: 1e-3
    weight_decay: 1e-6
    sched:
      name: NoamAnnealing
      warmup_steps: 4000
      min_lr: 1e-5
      d_model: ${model.d_model}

Also, the Cosine Annealing scheduler of hifigan seems to suffer from a similar issue. One can go to the scheduler and fiddle with the scalar as needed, but it would be much more efficient if things can be done via the config.

**Comment by titu1994** [comment]
Noam is a unique schedule. It's schedule is determined not based on a peak Lr and a min Lr but instead on the factors of number of steps, number of warmup steps, the sqrt of D_model (generally of a transformer) and the linear scalar of this schedule (here, optim.lr).

In literature as in most implementations of Noam, it is impossible to exactly determine and set a peak Lr and a min Lr simple because amy combination of these 4 factors and obtain exact same values of peak and min lrs. In stead, the number of warmup steps and the approximate peak is mentioned and it is upto the user to determine coefficients that will meet the criterion. 

Note that Noam is the only such unique optimizer in NeMo as of now, and all other optimizers okay the general concept of peak and min lrs exactly. Cosine annealing is one of them. The optim.lr is exactly what will be attained ad the peak Lr after a finite number of warmup steps, and the min Lr is exactly what will be obtained near the end of training.

**Comment by DalHyun** [comment]
Thank you Titu. I added a linear scalar to the FastSpeech2 scheduler, and it works fine.
As to Hifigan, automatic optimization is turned off at the initialization of the model.
You would have to step both optimizer AND scheduler for each training iteration, but currently it updates only the optimizer.
Hence the unchanging, extremely low learning rate.
Changes such as following might do.

self.manual_backward(loss_d)
self.optim_d.step()
# ADD-?
self.scheduler_d.step()
# ?-ADD 
...
self.optim_g.step()
# ADD-?
self.scheduler_g.step()
# ?-ADD

**Comment by titu1994** [comment]
Hmm that's a bug. @blisc could you take a look

**Comment by Oktai15** [comment]
@titu1994 @blisc I will fix it today

**Comment by blisc** [comment]
Can you try again with the main branch and let us know if this still occurs?
